File: 153-find-minimum-in-rotated-sorted-array/153-find-minimum-in-rotated-sorted-array.c


int findMin(int* nums, int numsSize){
    // Binary search
    int left = 0, right = numsSize - 1;
    while(left < right){
        int mid = (left + right) / 2;
        /*
          If nums[mid] > nums[right], it means this part of array is sorted already, and you won't find the solution here. 
        */
        if(nums[mid] > nums[right])
            left = mid + 1;
        else
            right = mid;
    }
    return nums[left];
}

File: 153-find-minimum-in-rotated-sorted-array/153-find-minimum-in-rotated-sorted-array.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        n = len(nums)
        
        low = 0
        high = n - 1
        
        minimum = nums[low]

        while low <= high:
            mid = (low + high) // 2
            if nums[mid] > nums[(mid + 1) % n]:
                minimum = nums[mid + 1]
                break
            elif nums[mid - 1] > nums[mid]:
                minimum = nums[mid]
                break
            elif nums[mid] < nums[high]:
                high = mid - 1
            else:
                low = mid + 1

        return minimum